### PR TITLE
Cross universe boundaries in ORANGE

### PR DESCRIPTION
**Pull request by elliottbiondo**
The PR adds the ability to track across universe boundaries in ORANGE: moving from parent to daughter, daughter to parent, or directly from daughter to daughter in the case where holes share a boundary. 

This PR also contains:
- minor refactor: moving `next_step_` and `next_surface_` from `OrangeTrackView` to `OrangeStateData`
- properly apply universe translations as specified in JSON

### CODE REVIEW

**Comment by elliottbiondo** [comment]
@sethrj draft status can now be removed; this is ready for review.

File: src/orange/OrangeTrackView.hh [hunk 1]
```diff
@@ -174,7 +172,7 @@ OrangeTrackView::OrangeTrackView(ParamsRef const& params,
     CELER_EXPECT(states_);
     CELER_EXPECT(track_slot_ < states.size());
 
-    CELER_ENSURE(!this->has_next_step());
+    states_.next_step[track_slot_] = 0;
```
Comment (sethrj) [inline src/orange/OrangeTrackView.hh]: We shouldn't be changing the underlying data in the construction of a view. If this is temporary data it should be local to the class; if it's persistent across kernels (or even with in different "views" in the same kernel) then it needs to be initialized in the `operator=`.

File: src/orange/OrangeTrackView.hh [hunk 2]
```diff
@@ -217,20 +215,39 @@ OrangeTrackView::operator=(Initializer_t const& init)
...
+        if (next_uid)
+        {
+            auto const& trans_id
+                = tracker.unit_record()
+                      .translations[vol_rec.daughter_translation.unchecked_get()];
+            auto const& trans = params_.translations[trans_id];
+
+            TranslatorDown td(trans);
+            auto new_pos = td(local.pos);
+            local.pos = new_pos;
```
Comment (sethrj) [inline src/orange/OrangeTrackView.hh]: It should be safe to do:
```suggestion
            local.pos = td(local.pos);
```
right?

We could also add an overload to TranslatorDown `void operator()(Real3*) const` (or even `Span<real_type, 3>`) that can do an in-place transformation: there will likely be times we want to do that.

File: src/orange/OrangeTrackView.hh [hunk 3]
```diff
@@ -454,15 +489,20 @@ CELER_FUNCTION void OrangeTrackView::move_to_boundary()
 CELER_FUNCTION void OrangeTrackView::move_internal(real_type dist)
 {
     CELER_EXPECT(this->has_next_step());
-    CELER_EXPECT(dist > 0 && dist <= next_step_);
-    CELER_EXPECT(dist != next_step_ || !next_surface_);
+    CELER_EXPECT(dist > 0 && dist <= states_.next_step[track_slot_]);
+    CELER_EXPECT(dist != states_.next_step[track_slot_]
+                 || !states_.next_surface[track_slot_]);
```
Comment (sethrj) [inline src/orange/OrangeTrackView.hh]: I see a lot of changes that could've been hidden had I made a private accessor `next_step` and `next_surface`. Could you replace `states_.next_step[track_slot_]` with `this->next_step()`? and `states_.next_surface[track_slot_]` with `this->next_surface()`?

File: src/orange/OrangeTrackView.hh [hunk 4]
```diff
@@ -454,15 +489,20 @@ CELER_FUNCTION void OrangeTrackView::move_to_boundary()
...
 
-    // Move and update next_step_
-    auto lsa = this->make_lsa();
-    axpy(dist, lsa.dir(), &lsa.pos());
+    // Move and update the next step
+    for (auto i : range(states_.level[track_slot_] + 1))
```
Comment (sethrj) [inline src/orange/OrangeTrackView.hh]: Probably want to do the same for `this->level()`.

File: src/orange/OrangeTrackView.hh [hunk 5]
```diff
@@ -492,47 +537,100 @@ CELER_FUNCTION void OrangeTrackView::cross_boundary()
...
 
-    lsa.surf() = init.surface.id();
-    lsa.sense() = init.surface.unchecked_sense();
+    // Starting with the current level (i.e., next_surface_level), iterate down
+    // into the deepest level
+    size_type current_level = sl.get();
+    auto current_uid = lsa.universe();
+    auto current_volume = tinit.volume;
+    VolumeRecord current_vol_rec
+        = params_.volume_records[tracker.unit_record().volumes[current_volume]];
```
Comment (sethrj) [inline src/orange/OrangeTrackView.hh]: To (possibly) reduce register pressure let's leave this as a const reference

File: src/orange/OrangeTrackView.hh [hunk 6]
```diff
@@ -492,47 +537,100 @@ CELER_FUNCTION void OrangeTrackView::cross_boundary()
...
+    {
+        // Get the translator at the parent level, in order to translate into
+        // daughter
+        current_trans
+            = params_.translations
+                  [this->make_tracker(current_uid)
+                       .unit_record()
+                       .translations[current_vol_rec.daughter_translation
+                                         .unchecked_get()]];
+        TranslatorDown current_translator(current_trans);
```
Comment (sethrj) [inline src/orange/OrangeTrackView.hh]: This is a little hard to follow, and there's no need for `current_trans` to be a local variable or in the parent scope. I think we'll also need the translations to be independent of the universe type, so we may need to store those on a separate ragged array...

Alternatively, and I think this is the way to go, is that we make the `Tracker` class responsible for giving us a translation/transformation for a certain volume. That kind of interface would allow us (in the future) to have some tracker types (SimpleUnit) return a translation from memory, and others (Array/DodeArray etc) do the translation on the fly, which can vastly reduce the memory burden for large pebble bed reactors.

File: src/orange/construct/OrangeInputIO.json.cc [hunk 7]
```diff
@@ -25,6 +25,21 @@
...
+namespace
+{
+
+//---------------------------------------------------------------------------//
+/*!
+ * Return the i-th translation from a flattened vector.
+ */
+celeritas::Translation make_translation(std::vector<double> const& translations,
+                                        celeritas::size_type i)
+{
```
Comment (sethrj) [inline src/orange/construct/OrangeInputIO.json.cc:38]: Maybe use a `Span<double>` and add some assertions for the size of the array

File: src/orange/OrangeTrackView.hh [hunk 8]
```diff
@@ -255,8 +272,14 @@ OrangeTrackView& OrangeTrackView::operator=(DetailedInitializer const& init)
...
-    states_.next_level[track_slot_]
-        = states_.next_level[init.other.track_slot_];
+    states_.surface_level[track_slot_]
+        = states_.surface_level[init.other.track_slot_];
+
+    states_.next_step[track_slot_] = states_.next_step[init.other.track_slot_];
+    states_.next_surface[track_slot_]
+        = states_.next_surface[init.other.track_slot_];
+    states_.next_surface_level[track_slot_]
+        = states_.next_surface_level[init.other.track_slot_];
```
Comment (sethrj) [inline src/orange/OrangeTrackView.hh]: If we're updating the direction (and immediately clearing the "next" information) these `next` assignment shouldn't be needed.

File: src/orange/OrangeTrackView.hh [hunk 9]
```diff
@@ -492,47 +537,100 @@ CELER_FUNCTION void OrangeTrackView::cross_boundary()
     CELER_EXPECT(this->is_on_boundary());
     CELER_EXPECT(!this->has_next_step());
 
-    auto lsa = this->make_lsa();
+    auto sl = states_.surface_level[track_slot_];
+    auto lsa = this->make_lsa(sl);
 
     if (CELER_UNLIKELY(lsa.boundary() == BoundaryResult::reentrant))
     {
+        CELER_ASSERT(sl.unchecked_get() == 0);
```
Comment (sethrj) [inline src/orange/OrangeTrackView.hh]: Would be clearer as `(sl == LevelId{0})`; and is this an intentional omission for now? When running through Celeritas we will definitely have reentrant boundaries that aren't at the global level (basically we might just get backscattered/deflected by a field when crossing an internal surface).

File: src/orange/OrangeTrackView.hh [hunk 10]
```diff
@@ -492,47 +537,100 @@ CELER_FUNCTION void OrangeTrackView::cross_boundary()
...
+
+        // Create local state on the daughter level
+        local.pos = current_translator(local.pos);
+        local.volume = {};
+        local.surface = {};
+        local.temp_sense = this->make_temp_sense();
+
+        current_volume = current_tracker.initialize(local).volume;
+        current_vol_rec = params_.volume_records[current_tracker.unit_record()
+                                                     .volumes[current_volume]];
```
Comment (sethrj) [inline src/orange/OrangeTrackView.hh]: Likewise, the tracker should abstract this.

File: src/orange/OrangeTrackView.hh [hunk 11]
```diff
@@ -492,47 +537,100 @@ CELER_FUNCTION void OrangeTrackView::cross_boundary()
...
+        current_lsa.boundary() = BoundaryResult::exiting;
+    }
 
-    // Reset boundary crossing state
-    lsa.boundary() = BoundaryResult::exiting;
+    states_.level[track_slot_] = LevelId{current_level};
+    states_.next_surface_level[track_slot_] = LevelId{};
 
     CELER_ENSURE(this->is_on_boundary());
+    this->clear_next_step();
```
Comment (sethrj) [inline src/orange/OrangeTrackView.hh:653]: Shouldn't this clear `next_surface_level`?

File: src/orange/OrangeTrackView.hh [hunk 12]
```diff
@@ -589,8 +687,10 @@ CELER_FUNCTION void OrangeTrackView::set_dir(Real3 const& newdir)
 CELER_FUNCTION void
 OrangeTrackView::find_next_step_impl(detail::Intersection isect)
 {
-    // Zero for top-level universe
+    // The UniverseID and LevelId corresponding to the level with with minium
+    // distance to intersection
     UniverseId min_uid{0};
+    LevelId min_level{0};
```
Comment (sethrj) [inline src/orange/OrangeTrackView.hh:798]: To reduce register usage/stack size we should avoid keeping both `min_uid` and `min_level`. After finding `min_level` we can determine min_uid (only needed if we intersect!) with `this->make_lsa(min_level).universe()`.

I'd also avoid initializing this to zero since the default value isn't used: better to let it default-initialize to an invalid (`bool(LevelId{}) == false`) state.

File: src/orange/OrangeTrackView.hh [hunk 13]
```diff
@@ -474,9 +514,14 @@ CELER_FUNCTION void OrangeTrackView::move_internal(real_type dist)
  */
 CELER_FUNCTION void OrangeTrackView::move_internal(Real3 const& pos)
 {
-    auto lsa = this->make_lsa();
-    lsa.pos() = pos;
-    lsa.surf() = LocalSurfaceId{};
+    for (auto i : range(states_.level[track_slot_] + 1))
+    {
+        auto lsa = this->make_lsa(LevelId{i});
+        lsa.pos() = pos;
```
Comment (sethrj) [inline src/orange/OrangeTrackView.hh]: `pos` here is global, and we need to propagate it down to each level. I suspect it would be "safer" (more robust against roundoff error if the deepest level is orders of magnitude displaced from the global level) to do that by updating each level by accumulating `pos - this->pos()`, but the simpler thing to do would be doing repeated global-to-local transformations like we do with the initialize function.

**Review by sethrj — COMMENTED** [review]
I think it looks pretty good, but we might want to pause for a second (either on this PR or the next) and think about how we can abstract or pull out some of the code into more testable classes. It's getting harder to tell what conditions we've thoroughly tested and what we haven't...

File: src/orange/OrangeTrackView.hh (hunk 9 above)
Comment (elliottbiondo) [reply src/orange/OrangeTrackView.hh]: Yes this was intentional. I am not sure I understand the use-case. Is this something specific to tracking through a B-field?

File: src/orange/OrangeTrackView.hh (hunk 2 above)
Comment (elliottbiondo) [reply src/orange/OrangeTrackView.hh]: (artifact of a refactor)

File: src/orange/OrangeTrackView.hh (hunk 9 above)
Comment (sethrj) [reply src/orange/OrangeTrackView.hh]: Yes, it's something you only see with charged particles and not in neutron/gamma transport. Basically, a step can be geometry-limited (and move to the boundary), but in the process of moving that step, the magnetic field could change its direction so that instead of exiting it heads back into the cell it was about to leave. Similarly, "multiple scattering" happens along every step for charged particles, and that changes the direction at the end of the step. So a track that's "on the boundary" about to leave the cell ends up bouncing back in without actually leaving.

Comment (sethrj) [reply src/orange/OrangeTrackView.hh]: If this is deliberately not implemented rather than an assertion for a normal code condition, also write like this:
```suggestion
        if (sl != LevelId{0})
        {
            CELER_NOT_IMPLEMENTED("reentrant surfaces with multilevel ORANGE geometry");
        }
```

File: src/orange/OrangeTrackView.hh (hunk 3 above)
Comment (elliottbiondo) [reply src/orange/OrangeTrackView.hh]: I'll do this for all of the fields of length `num_tracks` on the state, i.e. those not accessed by LSA.

File: src/orange/OrangeTrackView.hh (hunk 5 above)
Comment (elliottbiondo) [reply src/orange/OrangeTrackView.hh]: This actually has to be `VolumeRecord const*` because it changes within the loop.

File: src/orange/OrangeTrackView.hh (hunk 11 above)
Comment (elliottbiondo) [reply src/orange/OrangeTrackView.hh:653]: Not sure what you mean? It does, and I think is should.

Comment (sethrj) [reply src/orange/OrangeTrackView.hh:653]: Sorry, the highlight isn't showing the context: a few lines earlier you explicitly set:
```
states_.next_surface_level[track_slot_] = LevelId{};
```
which I think can be deleted.

File: src/orange/OrangeTrackView.hh (hunk 12 above)
Comment (elliottbiondo) [reply src/orange/OrangeTrackView.hh:798]: It actually does need to be initialized to zero because the argument `isect` always corresponds to level 0.

File: src/orange/OrangeTrackView.hh (hunk 1 above)
Comment (elliottbiondo) [reply src/orange/OrangeTrackView.hh]: The `CELER_ENSURE` I removed calls:
```
CELER_FUNCTION bool OrangeTrackView::has_next_step() const
{
    return states_.next_step[track_slot_] != 0;
}
```
which fails if states_.next_step is initialized. Is there a way of initialing all the values of a collection to 0?

**Comment by elliottbiondo** [comment]
@sethrj this is ready for another look

File: src/orange/OrangeData.hh [hunk 14]
```diff
@@ -58,8 +59,7 @@ struct VolumeRecord
 
     logic_int max_intersections{0};
     logic_int flags{0};
-    UniverseId daughter;
-    TranslationId daughter_translation;
+    DaughterId daughter_id;
```
Comment (sethrj) [inline src/orange/OrangeData.hh:62]: This and below could probably be replaced with an `ItemMap<LocalVolumeId, DaughterId>`: since the IDs are small we could probably just store one for every volume in the unit as part of the `SimpleUnitRecord`..

File: src/orange/OrangeTrackView.hh [hunk 15]
```diff
@@ -217,20 +248,31 @@ OrangeTrackView::operator=(Initializer_t const& init)
...
         lsa.boundary() = BoundaryResult::exiting;
 
-        auto const& vol_rec = tracker.unit_record().volumes[tinit.volume];
-        next_uid = params_.volume_records[vol_rec].daughter;
-        ++level;
+        next_uid = tracker.daughter_uid(tinit.volume);
+
+        if (next_uid)
+        {
+            auto const& trans = tracker.translation(tinit.volume);
```
Comment (sethrj) [inline src/orange/OrangeTrackView.hh]: I was thinking along the lines of:
```suggestion
        daughter_id = tracker.daughter(tinit.volume);

        if (daughter_id)
        {
            TranslationId trans_id = params_.daughters[daughter_id];
            auto const& trans = params_.translations[trans_id];
```
since we might end up repeating the same code for the other universe types. Thoughts?

File: src/orange/OrangeTypes.hh [hunk 16]
```diff
@@ -48,6 +54,16 @@ using UniverseId = OpaqueId<struct Universe>;
...
 
+//! Data to specify an embeded universe withing a volume
+struct Daughter
+{
+    UniverseId universe_id;
+    TranslationId translation_id;
+};
+
+//! Opaque index for "simple unit" data
+using DaughterId = OpaqueId<struct Daughter>;
```
Comment (sethrj) [inline src/orange/OrangeTypes.hh]: As with `SimpleUnitRecord`, you can rely on the `struct` inside the template parameter to forward-declare the class, and move the `Daughter` definition down with the other structs. Could you then also keep the IDs here alphabetically sorted?

File: src/orange/OrangeData.hh (hunk 14 above)
Comment (elliottbiondo) [reply src/orange/OrangeData.hh:62]: Every volume in a unit has a VolumeRecord, so that makes sense as an ItemMap. However, the vast majority of volumes will not have daughters, so it seems strange to store a DaughterId for every volume, even if it is small. Since daughters access will all be handled in one place in the SimpleUnitTracker (per your API below), the syntactic benefit doesn't seem worth it to me.

File: src/orange/OrangeTrackView.hh (hunk 15 above)
Comment (elliottbiondo) [reply src/orange/OrangeTrackView.hh]: I agree that this is better. The only downside is it locks us into daughters being stored the same way for array units, which they probably will be, but I haven't thought about it enough. I will change to this.

File: src/orange/OrangeTrackView.hh (hunk 1 above)
Comment (elliottbiondo) [reply src/orange/OrangeTrackView.hh]: @sethrj I think you may have missed this comment, and another below on your second look

File: src/orange/OrangeData.hh (hunk 14 above)
Comment (sethrj) [reply src/orange/OrangeData.hh:62]: Either way we're storing a daughter ID for every volume: here it's just embedded into the size of the volume record. The only away to avoid that extra storage is to have a map (hash table, associative vector, etc.) that stores volume ID and daughter ID. I don't have a strong opinion about whether daughter ID is part of the volume record or the unit storage, but it seems redundant to have both.

Comment (elliottbiondo) [reply src/orange/OrangeData.hh:62]: Okay now I see what you are saying. There is no reason to store the ItemRange of Daughters on the UnitRecord. I only did it that way to mirror what is done with the other data within UnitInserter. I fixed it so that now DaughterId is only stored on VolumeRecord.

**Comment by elliottbiondo** [comment]
@sethrj this is ready for another look

**Review by sethrj — APPROVED** [review]
Looks good! Thanks for all the revisions.
